android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt

```diff
@@ -0,0 +1,89 @@
+import host.exp.exponent.ABIVersion
+import host.exp.exponent.Constants
+import host.exp.exponent.analytics.EXL
+import host.exp.exponent.kernel.ExperienceId
+import host.exp.exponent.notifications.PushNotificationHelper
+import host.exp.exponent.notifications.model.ScopedNotificationRequest
+import host.exp.exponent.storage.ExponentDB
+import org.json.JSONException
+import org.json.JSONObject
+
+class ExpoFirebaseMessagingDelegate(context: Context) : FirebaseMessagingDelegate(context) {
+  override fun onNewToken(token: String) {
+    if (!Constants.FCM_ENABLED) {
+      return
+    }
+    super.onNewToken(token)
+    FcmRegistrationIntentService.registerForeground(context.applicationContext, token)
+  }
+
+  override fun onMessageReceived(remoteMessage: RemoteMessage) {
+    if (!Constants.FCM_ENABLED) {
+      return
+    }
+    val experienceDBObject = ExponentDB.experienceIdToExperienceSync(remoteMessage.data["experienceId"])
+    if (experienceDBObject != null) {
+      try {
+        val manifest = JSONObject(experienceDBObject.manifest)
+        val sdkVersion = ABIVersion.toNumber(manifest.getString("sdkVersion")) / 10000
+
+        // If an experience is on SDK newer than 39, that is SDK40 and beyond up till UNVERSIONED
+        // we only use the new notifications API as it is going to be removed from SDK40.
+        if (sdkVersion >= 40) {
+          dispatchToNextNotificationModule(remoteMessage)
+          return
+        } else if (sdkVersion == 38 || sdkVersion == 39) {
+          // In SDK38 and 39 we want to let people decide which notifications API to use,
+          // the next or the legacy one.
+          val androidSection = manifest.optJSONObject("android")
+          if (androidSection != null) {
+            val useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false)
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(remoteMessage)
+              return
+            }
+          }
+        }
+        // If it's an older experience or useNextNotificationsApi is set to false, let's use the legacy notifications API
+        dispatchToLegacyNotificationModule(remoteMessage)
+      } catch (e: JSONException) {
+        e.printStackTrace()
+        EXL.e("expo-notifications", "Couldn't parse the manifest.")
+      }
+    } else {
+      EXL.e("expo-notifications", "No experience found for id ${remoteMessage.data["experienceId"]}")
+    }
+  }
+
+  private fun dispatchToNextNotificationModule(remoteMessage: RemoteMessage) {
+    super.onMessageReceived(remoteMessage)
+  }
+
+  private fun dispatchToLegacyNotificationModule(remoteMessage: RemoteMessage) {
+    PushNotificationHelper.getInstance().onMessageReceived(
+      context,
+      remoteMessage.data["experienceId"],
+      remoteMessage.data["channelId"],
+      remoteMessage.data["message"],
+      remoteMessage.data["body"],
+      remoteMessage.data["title"],
+      remoteMessage.data["categoryId"]
```

Comment:
What would you say if we wrap this `RemoteMessage.data` into our custom data class? 🤔 
I saw that this `RemoteMessage.data` is just a [`Map<string, string>`](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage#getData()) and we're accessing every here and there some sophisticated keys in this data payload. I image we can abstract this `Map<string, string>` into some generic `DataClass` with some known properties & provide some inline fallback mechanisms for missing values? 🤔

Reply:
If you looked closely you'd notice that this method is marked as _legacy_, in `expo-notifications` we already do it this way — check out `JSONNotificationContentBuilder`!

https://github.com/expo/expo/blob/f36a5161b98f879defd331133604c2677a87b253/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/JSONNotificationContentBuilder.java#L37-L61